rts/lra/lra-cdi-rest/src/main/java/org/jboss/narayana/rts/lra/cdi/LraAnnotationProcessingExtension.java

```diff
@@ -0,0 +1,149 @@
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.narayana.rts.lra.cdi;
+
+import java.lang.annotation.Annotation;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.AnnotatedMethod;
+import javax.enterprise.inject.spi.DeploymentException;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.WithAnnotations;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+import org.jboss.narayana.rts.lra.annotation.Compensate;
+import org.jboss.narayana.rts.lra.annotation.Complete;
+import org.jboss.narayana.rts.lra.annotation.LRA;
+import org.jboss.narayana.rts.lra.annotation.Status;
+
+
+/**
+ * <p>
+ * CDI extension working in concord of LRA filters.
+ * <p>
+ * When added at the class path of the project this extension validates
+ * if the classes contain compulsory annotation complementary to {@link LRA}.
+ * The rules of what are compulsory annotations and their attributes
+ * are defined in LRA specification.
+ * <p>
+ * In case of validity violation the {@link DeploymentException} is thrown.
+ * 
+ * @author Ondra Chaloupka <ochaloup@redhat.com>
+ */
+public class LraAnnotationProcessingExtension implements Extension {
+
+    <X> void processLraAnnotatedType(@Observes @WithAnnotations({LRA.class}) ProcessAnnotatedType<X> classAnnotatedWithLra) {
+
+        // All compulsory LRA annotations are available at the class
+        Supplier<Stream<AnnotatedMethod<? super X>>> sup = () -> classAnnotatedWithLra.getAnnotatedType().getMethods().stream();
+        Set<Class<? extends Annotation>> missing = new HashSet<>();
+        if(!sup.get().anyMatch(m -> m.isAnnotationPresent(Compensate.class))) missing.add(Compensate.class);
+        if(!sup.get().anyMatch(m -> m.isAnnotationPresent(Complete.class))) missing.add(Complete.class);
+        if(!sup.get().anyMatch(m -> m.isAnnotationPresent(Status.class))) missing.add(Status.class);
```

Comment:
Won't the other checks that you coded on lines 95, 98 and 101 catch the missing annotations.

Reply:
Yes, it could match the trouble too but my point was to provide as most comprehensible description of the exception. Thus I split it and possibly duplicate the check. But the user gets, I hope, more precise message to understand the issue.